File: julia/Euler/src/bin/p0061.jl

# project euler: problem 61

module Prob0061

function make_polynum_tbl(max_nsides_polygon)
    tbl = Dict{Int, Dict{Int, Vector{Int}}}()
    for i = 3:max_nsides_polygon
        x = Dict{Int, Vector{Int}}()
        acc = 0
        step = i - 2
        j = 1
        while true
            acc += j
            if acc >= 10_000
                break
            end
            if acc >= 1_000 && acc % 100 > 9
                k, v = divrem(acc, 100)
                x[k] = push!(get(x, k, []), v)
            end
            j += step
        end
        tbl[i] = x
    end
    tbl
end

function find_closed_paths(max_nsides_polygon)
    closed_paths::Vector{Vector{Int}} = []
    tbl = make_polynum_tbl(max_nsides_polygon)
    stop_condition = (1 << (max_nsides_polygon + 1)) - 8

    function get_next_states((bits, path))
        next_states::Vector{Tuple{Int, Vector{Int}}} = []

        # bits: (when max_nsides_polygon = 8)
        #   0b######000
        #     ||||||
        #     |||||+- triangle
        #     ||||+-- square
        #     |||+--- pentagonal
        #     ||+---- hexagonal
        #     |+----- heptagonal
        #     +------ octagonal
        if bits == stop_condition
            if path[1] == path[end]
                # Found a closed path
                push!(closed_paths, path)
            end
        else
            for nsides = 3:(max_nsides_polygon - 1)
                p_bit = 1 << nsides
                if bits & p_bit != 0
                    continue
                end
                if haskey(tbl[nsides], path[end])
                    for x in tbl[nsides][path[end]]
                        push!(next_states, (bits | p_bit, push!(copy(path), x)))
                    end
                end
            end
        end
        next_states
    end

    # Search for all closed paths
    q::Vector{Tuple{Int, Vector{Int}}} = []
    for (k, vs) in tbl[max_nsides_polygon]
        for v in vs
            push!(q, (1 << max_nsides_polygon, [k, v]))
        end
    end
    while length(q) > 0
        state = pop!(q)
        for next_state in get_next_states(state)
            push!(q, next_state)
        end
    end

    closed_paths
end

function solve_0061(max_nsides_polygon::Int = 8)
    function is_distinct_numbers(lst)
        tmp = Set{Int}()
        for i = 1:(length(lst) - 1)
            push!(tmp, lst[i] * 100 + lst[i + 1])
        end
        length(tmp) == (length(lst) - 1)
    end

    @assert (max_nsides_polygon > 3) "invalid parameter"

    cycles::Vector{Vector{Int}} = []
    for path in find_closed_paths(max_nsides_polygon)
        # All numbers in a cycle are different from each other's
        if is_distinct_numbers(path)
            push!(cycles, path)
        end
    end

    # There exists only one cycle
    if length(cycles) == 1
        return sum(cycles[1][2:end]) * 101
    end
    @assert false "not reached"
end

end #module

using .Prob0061: solve_0061
export solve_0061

File: typescript/euler/bin/p0061.ts
// project euler: problem 61

import { sum } from "../lib/math.ts";
import { range } from "../lib/util.ts";

const makePolygonalTbl = (
  maxNumSidesPolygon: number,
): Map<number, Map<number, number[]>> => {
  const tbl = new Map<number, Map<number, number[]>>();
  for (const i of range(3, maxNumSidesPolygon + 1)) {
    const mapObj = new Map<number, number[]>();
    const step = i - 2;
    let acc = 0;
    let j = 1;
    while (true) {
      acc += j;
      if (acc >= 10_000) {
        break;
      }
      if (acc >= 1_000 && acc % 100 >= 10) {
        const key = Math.trunc(acc / 100);
        const val = acc % 100;
        mapObj.set(key, (mapObj.get(key) || []).concat([val]));
      }
      j += step;
    }
    tbl.set(i, mapObj);
  }

  return tbl;
};

const findClosedPaths = (maxNumSidesPolygon: number): number[][] => {
  const paths: number[][] = [];
  const polyTbl = makePolygonalTbl(maxNumSidesPolygon);

  // example: (when maxNumSidesPolygon = 8)
  //   0b######000
  //     ||||||
  //     |||||+- triangle
  //     ||||+-- square
  //     |||+--- pentagonal
  //     ||+---- hexagonal
  //     |+----- heptagonal
  //     +------ octagonal
  const stopCondition = (1 << (maxNumSidesPolygon + 1)) - 8;

  const getNextState = (state: [number, number[]]): [number, number[]][] => {
    const states: [number, number[]][] = [];
    const bits = state[0];
    const path = state[1];
    if (bits == stopCondition && path.at(0)! == path.at(-1)!) {
      paths.push(path);
    } else {
      for (const i of range(3, maxNumSidesPolygon)) {
        const p_bit = 1 << i;
        if ((bits & p_bit) != 0) {
          continue;
        }
        const next_tbl = polyTbl.get(i)!;
        if (next_tbl.has(path.at(-1)!)) {
          for (const x of next_tbl.get(path.at(-1)!)!) {
            states.push([bits | p_bit, [...path, x]]);
          }
        }
      }
    }

    return states;
  };

  // search by DFS
  const q: [number, number[]][] = [];
  for (const [k, vs] of polyTbl.get(maxNumSidesPolygon)!.entries()) {
    for (const v of vs) {
      q.push([1 << maxNumSidesPolygon, [k, v]]);
    }
  }
  while (q.length > 0) {
    //console.log(q);
    const state = q.pop()!;
    for (const next_state of getNextState(state)) {
      q.push(next_state);
    }
  }

  return paths;
};

export const compute = (maxNumSidesPolygon: number): string => {
  const isDistinctNumbers = (lst: number[]): boolean => {
    const tmp: Set<number> = new Set();
    for (const i of range(0, lst.length - 1)) {
      tmp.add(lst[i] * 100 + lst[i + 1]);
    }
    return tmp.size == lst.length - 1;
  };

  const cycles: number[][] = [];
  for (const path of findClosedPaths(maxNumSidesPolygon)) {
    // All numbers in a cycle are different from each other's
    if (isDistinctNumbers(path)) {
      cycles.push(path);
    }
  }

  // There exists only one cycle
  if (cycles.length == 1) {
    return String(sum(cycles[0].slice(1)) * 101);
  }

  throw new Error("unreachable");
};

export const solve = (): string => compute(8);
